includes/mailpoet/class-sensei-mailpoet.php

```diff
@@ -0,0 +1,385 @@
+	 *
+	 * @param int  $user_id User ID.
+	 * @param int  $course_id Post ID.
+	 * @param bool $is_enrolled Enrollment status.
+	 *
+	 * @access private
+	 * @return void
+	 */
+	public function maybe_add_student_course_subscriber( $user_id, $course_id, $is_enrolled ) {
+		if ( $is_enrolled ) {
+			$this->add_student_subscriber( $course_id, $user_id );
+		} else {
+			$this->remove_student_subscribers( $course_id, array( $user_id ) );
+		}
+	}
+
+	/**
+	 * Remove student as subscriber to a course list on MailPoet. A proxy to SenseiMailPoet::remove_student_subscribers.
+	 *
+	 * @param int $user_id User ID.
+	 * @param int $course_id Post ID.
+	 *
+	 * @access private
+	 * @return void
+	 */
+	public function remove_student_course_subscriber( $user_id, $course_id ) {
+		$this->remove_student_subscribers( $course_id, array( $user_id ) );
+	}
+
+	/**
+	 * Add student as subscriber to a course list on MailPoet. A proxy to SenseiMailPoet::add_student_subscriber.
+	 *
+	 * @param int $user_id User ID.
+	 * @param int $course_id Post ID.
+	 *
+	 * @access private
+	 * @return void
+	 */
+	public function add_student_course_subscriber( $user_id, $course_id ) {
+		$this->add_student_subscriber( $course_id, $user_id );
+	}
+
+	/**
+	 * Get subscribers of a MailPoet list.
+	 * Returns null if the list does not exist.
+	 *
+	 * @since $$next-version$$
+	 * @param int $mp_list_id MailPoet list ID.
+	 *
+	 * @return array
+	 */
+	public function get_mailpoet_subscribers( $mp_list_id ) {
+		try {
+			return $this->mailpoet_api->getSubscribers( array( 'listId' => (int) $mp_list_id ) );
+		} catch ( \MailPoet\API\MP\v1\APIException $exception ) {
+			return array();
+		}
+	}
+
+	/**
+	 * Figure out which subscribers to add and remove from lists on MailPoet for courses/groups.
+	 *
+	 * @since $$next-version$$
+	 * @param array      $students A list of students belonging to this group/course list.
+	 * @param array      $subscribers A list of subscribers already on MailPoet.
+	 * @param string|int $list_id ID of the list on MailPoet.
+	 *
+	 * @return void
+	 */
+	public function sync_subscribers( $students, $subscribers, $list_id ) {
+		$options = array(
+			'send_confirmation_email'      => false,
+			'schedule_welcome_email'       => false,
+			'skip_subscriber_notification' => true,
+		);
+		foreach ( $students as $student ) {
+			if ( ! array_key_exists( $student['email'], $subscribers ) ) {
+				$subscriber_data = array(
+					'email'      => $student['email'],
+					'first_name' => $student['display_name'],
+				);
+
+				try {
+					$mp_subscriber = $this->mailpoet_api->getSubscriber( $student['email'] );
+					$this->mailpoet_api->subscribeToList( $mp_subscriber['id'], $list_id, $options );
+				} catch ( \MailPoet\API\MP\v1\APIException $exception ) {
+					if ( 4 === $exception->getCode() ) {
```

Comment:
TBH the subscriber should always exist since MailPoet adds all WordPress users as subscribers by default. I was just getting some IDE warnings about not catching exceptions. I could use `empty()` or `isset()` to check too.